README
                               pcsets:
                     Pitch Class Sets for Python

                            version 2.0.2

                           Bruce H. McCosar


Welcome, and thank you for downloading pcsets. This is the third
official (non-beta) release. With this release, I start the pcsets
tutorial see "About this Package", below).

Learning about Pitch Class Sets is easy -- the four modules presented
here are, like many Python programs, self documenting. They work well
with pydoc <http://docs.python.org/lib/module-pydoc.html>.

Learning how to use these for practical compositional purposes, on the
other hand ... well, that takes a lot of individual experimentation. I
have started a short tutorial, which will eventually demonstrate some
of the techniques I've learned. However, you are a much better judge
of what you want to know than I -- please, try the module out, and
most of all, have fun learning something new.


                        About Pitch Class Sets
                        ======================

Pitch Class Sets are a mathematical model for analyzing and composing
music. Each note 'C' through 'B' has an equivalent pitch class number
0 through 11. Sets of these numbers may be operated on by mathematical
functions such as transpose and invert.

The goal of this project is to, eventually, have:

  * A Python library capable of fully implementing Pitch Class Sets
    and their common operations, as well as several convenience
    functions to bring these abstract concepts to the real world.
    (Mapping pitch classes to note names, for instance).

  * A tool for composition.  Some applications are harmonization,
    chord voicing generation, and melodic motif creation.

  * More exotic goals -- creation of new chordal elements, musical
    progressions, and harmonic relationships.


                          About this Package
                          ==================

The API will not change for any of the modules referred to as 'core'.
Bugs will be fixed and new modules will be added, but you won't wake
up one day and find that PcSets wants input in statcoulombs or Dutch
Guldens. In the version 2 series, the core will always behave like
the core. Occasionally new functionality will be introduced, but the
original functionality will never be 'broken' -- that's why I spent all
that time writing unit tests.


THE CORE

This is the core as of 2.0.0:


    pcsets.pcset

        The base class, PcSet, includes methods that operate on single
	sets, such as inversion and transposition.

    pcsets.pcops

        Operations on two or more sets, such as subset_of(a,b).

    pcsets.catalog

        Generates the entire catalog of 224 prime sets as a Python
        object. Since this takes a while to generate, it saves the
        catalog in a pickle file (catalog.pkl) for future use.

    pcsets.noteops

        The 'universal translator' from PcSets to named notes and
        vice versa.


Any *new* modules will enter as 'experimental', however, not core.
Experimental modules can change at any time.


EXPERIMENTAL MODULES

With version 2.0.1, I introduced the first experimental module:


    pcsets.tonerow

        Implements the ToneRow class.  Unlike PcSets, which are
	unordered, a ToneRow consists of all 12 pitches in an
	*ordered* arrangement.


There is a lot of good information on this subject in the Straus book
referenced below. I've also put a lot of time into writing documentation
strings for the module; a run through it with pydoc will probably tell
you somewhere between too much and far too much ;-)

Another experimental module in development (and still in hiding,
until I settle on a usable, workable interface):


  * Operations on the familiar chords and scales -- sort of a noteops
    for the common language of chord-scale theory.


These modules won't make it to the core until at least version 2.1.


THE TUTORIAL

Finally, I've made a start on the tutorial.  My hope is that working
through the documentation -- even for features that haven't been added
yet! -- will let me streamline the API and introduce more relevant
features in future releases.


Be sure to check back at the project website for updates.  I seem to
have embraced Google Code's slogan, "Release early, release often."


                   http://code.google.com/p/pcsets/



--BMC


                              Versioning
                              ==========


In an official release, the version numbers A.B.C promise the following:


    A = Major Number.  None of the modules referred to as 'core' will
        change except to correct errors. Functionality may be added,
        but none of the existing methods, classes, or functions
        will be altered in terms of their expected input or output
        characteristics.

        Later 'experimental' modules may become part of the core, but
        modules will never leave the core.

    B = Minor Number.  Previously released, tested, and proven
      	experimental modules may be moved into 'core'.  New abilities
	may be added to core modules, but no existing core interface
	will change.

    C = Patch Number.  New modules may be added as 'experimental'.
      	Also incremented for bug fixes and added documentation,
	including example code and tests.


[Note: this scheme had to be corrected from the version 2.0.1 release;
see the ChangeLog for details.]


                           About the Author
                           ================

Bruce H. McCosar is a middle and high school science teacher residing
in Gainesville, Florida; he switched to teaching after a career in
medicinal chemistry. He is a member of the Mvskoke (Creek) Nation.

Besides a lifelong interest in science, Bruce is a musician. He plays
bass guitar, electric guitar, Hammond organ, and conga drums.

In 2006, he began releasing his music online under a Creative Commons
license for free download at Jamendo. So far he has two albums,
'evolution' (Nov. 2006) and 'handmade' (Feb. 2007), available from his
artist page at:


          http://www.jamendo.com/us/artist/bruce.h.mccosar/


And if all that wasn't enough, he then got into Python programming!


                              References
                              ==========

Web

 * If you want a short introduction or tutorial, Jay Tomlin's site is
   the best. Sort of the 'Classics Illustrated' of Pc theory.

          http://www.jaytomlin.com/music/settheory/help.html


 * If you want a LOT of information without chasing down a book,
   well... here's the next best thing.

                http://solomonsmusic.net/setheory.htm


Text

  * The classic of the field, "The Structure of Atonal Music", by
    Allen Forte (1973).

  * A relatively new (but extremely thorough and readable) work,
    "Introduction to Post-Tonal Theory", by Joseph Straus (3rd ed.,
    2005).



# pcsets 2.0.2 -- Pitch Class Sets for Python.
#
# Copyright 2007 Bruce H. McCosar
#
# This file is part of the package 'pcsets'
#
# The package 'pcsets' is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# The package 'pcsets' is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty
# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# $Id: README 209 2007-08-19 16:41:44Z mccosar $


ChangeLog
ChangeLog: pcsets version 2 series                            -*-Text-*-
========================================================================

  PCSETS VERSION 2.0.2

  My apologies for the lag in updating this module. I'm a middle school
  science teacher by day, a musician by night, and ... well, evidently
  a programmer in the only thing left -- The Twilight Zone(tm). School
  started up again, and I'm in a constant battle to get things done.

  First, I had to fix the versioning information in the README file.
  The one that was published in the beta versions -- and all the way
  up to version 2.0.1 -- was, I guess you could say, a missed merge.
  All this time, I thought the text read correctly.  I focused on the
  'big picture' items at the top of the file, and sort of assumed I'd
  merged that change from the now-extinct 'Monolith' branch. [One with
  pcops integrated into PcSets... bad idea for various reasons, so it
  went to visit the Triceratops early on.]

  Next, I started a tutorial.  Right now, it's only up to Chapter 2.
  However, the finished product will have an index, references, and
  some handy diagrams.

  Finally, I upgraded pcsets.tonerow with a new function and a new
  method:

    pcsets
    `-- tonerow
        |-- ToneRow
        |   `-- contour
        `-- randomrow

  I put in a decent amount of documentation for these additions; have
  a look at the module with pydoc and it should provide a sort of
  introduction. Until I get to that section of the tutorial, that is
  ;-)

  svn_loc: atonal/tags/pcsets-2.0.2
  product: pcsets-2.0.2.tar.gz
  webpage: http://code.google.com/p/pcsets/
  dl_link: http://pcsets.googlecode.com/files/pcsets-2.0.2.tar.gz


------------------------------------------------------------------------
r188 | mccosar | 2007-08-06 22:22:15 -0400 (Mon, 06 Aug 2007) | 16 lines

  PCSETS VERSION 2.0.1

  Presenting the new module, pcsets.tonerow, which implements 12-tone
  rows and the standard group operations on them (P, R, I, and RI).
  This is an *experimental* module, and won't be in final form until
  at least version 2.1; I need to add a lot of features, such as a
  generator function (in the form f(pcs) -> ToneRow).

  In the meantime, the module is heavily populated with documentation
  strings; a run through it with pydoc might be a good introduction.

  svn_loc: atonal/tags/pcsets-2.0.1
  product: pcsets-2.0.1.tar.gz
  webpage: http://code.google.com/p/pcsets/
  dl_link: http://pcsets.googlecode.com/files/pcsets-2.0.1.tar.gz


------------------------------------------------------------------------
r176 | mccosar | 2007-07-27 23:06:52 -0400 (Fri, 27 Jul 2007)

  PCSETS VERSION 2.0.0 OFFICIAL RELEASE


  pcsets/*.py: Docstrings were updated to reflect current status; they
  were also reformatted, if necessary, to look better with pydoc.

  pcsets.catalog: Somehow this got by with only rudimentary
  docstrings. I added them. Actually, I added a lot -- the module doc
  is about the size of a mini tutorial. What can I say, the subject
  fascinates me.

  pcsets.catalog: Changed the names of methods rewrite, rebuild, and
  retrieve to the same names with an underscore '_' preface. These are
  more like internal functions.

  ./runtest.py: Added a notice that only shows up for the 'catalog'
  tests -- "Generating prime set catalog... (this may take a
  moment)" The same message was inserted at the start of the
  pcsets.catalog.showcatalog() demo function.

  pcsets.pcops.OpSet: now returns the string 'None' when __str__
  output is requested, but no solution sets were found.


  svn_loc: atonal/tags/pcsets-2.0.0
  product: pcsets-2.0.0.tar.gz
  webpage: http://code.google.com/p/pcsets/
  dl_link: http://pcsets.googlecode.com/files/pcsets-2.0.0.tar.gz


------------------------------------------------------------------------
r161 | mccosar | 2007-07-23 19:58:12 -0400 (Mon, 23 Jul 2007)

  PCSETS VERSION 2.0.0 BETA 3 RELEASE


  Additions to Core API (see full list in the previous ChangeLog entry)

  pcsets/
  |-- pcset
  |   `-- PcSets
  |       `-- reduced +    # These additions are
  `-- pcops                # all shortcut functions
      |-- rel_Tn +         # for otherwise 'wordy'
      `-- rel_TnI +        # operation sequences.


  1. Two new test types were added to test/test_noteops, both having to
     do with scales. The 'minimum conflict' setting should be expected
     to reproduce the common 'spellings' of all the possible major
     scales from B major to Db major. The only expected error would
     be the enharmonic F# and Gb scales, which would require Cb or E#
     (illegal, since in pcsets, natural notes are always natural notes).

     Also, it is expected that common alternative scales such as
     the melodic minor and harmonic minor -- which can be represented
     by an example scale with a single accidental -- should be able
     to reproduce that form.  For example, the harmonic minor is often
     given as a modification of the A minor scale to "A B C D E F G#"

  2. By default, pcsets.pcops.OpSet is not a *-exported class. Instead,
     four functions in pcsets.pcops return their solution sets as
     OpSet instances. However, in playing around with the system in
     interactive mode, it was found a __str__ function was VERY helpful
     for quick analysis:

       >>> from pcsets.pcset import *
       >>> from pcsets.pcops import *
       >>> pcs = PcSet('015')
       >>> cscale = PcSet('024579B')
       >>> print fit_in(cscale,pcs)   # no need to catch the object
       T(4) T(11) TnI(0) TnI(5)
       >>> print pcs.T(4)             # should be a subset of cscale
       459

     Two new tests were added to test/test_pcops to ensure this ability
     doesn't break if changes are made to the package internals.


------------------------------------------------------------------------
r138 | mccosar | 2007-07-19 20:32:56 -0400 (Thu, 19 Jul 2007) | 42 lines

  PCSETS VERSION 2.0.0 BETA 2 RELEASE


test/__init__.py
#!/usr/bin/env python

# pcsets 2.0.2 -- Pitch Class Sets for Python.
#
# Copyright 2007 Bruce H. McCosar
#
# This file is part of the package 'pcsets'
#
# The package 'pcsets' is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# The package 'pcsets' is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty
# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# $Id: __init__.py 191 2007-08-07 03:15:46Z mccosar $

__all__ = """
test_catalog
test_noteops
test_pcops
test_pcset
test_tonerow
""".split()


runtest.py
#!/usr/bin/env python

# pcsets 2.0.2 -- Pitch Class Sets for Python.
#
# Copyright 2007 Bruce H. McCosar
#
# This file is part of the package 'pcsets'
#
# The package 'pcsets' is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# The package 'pcsets' is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty
# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# $Id: runtest.py 191 2007-08-07 03:15:46Z mccosar $

"""
Test runner docs will go here.  Basically, you can specify
which module to test from the command line --

python runtest.py        # tests them all

python runtest.py pcset  # just tests module pcset
"""

from __future__ import print_function
import sys
import unittest

alltests = """
pcset
pcops
catalog
noteops
tonerow
""".split()


class NoSuchTestError(Exception):
    pass


def match(name):
    if name not in alltests:
        raise NoSuchTestError("Test '%s' not found." % name)
    else:
        return name


def run(testlist, level=2):
    for name in testlist:
        if name:
            print("TESTING MODULE: pcsets.%s" % name)
            if name == 'catalog':
                print("Generating prime set catalog...",
                      "(this may take a moment)")
            test = 'test.test_' + name
            suite = unittest.TestLoader().loadTestsFromName(test)
            unittest.TextTestRunner(verbosity=level).run(suite)

if __name__ == '__main__':
    args = sys.argv[1:]
    if args:
        tests = [match(name) for name in args]
        run(tests)
    else:
        run(alltests, level=1)


pcsets/__init__.py
#!/usr/bin/env python

# pcsets 2.0.2 -- Pitch Class Sets for Python.
#
# Copyright 2007 Bruce H. McCosar
#
# This file is part of the package 'pcsets'
#
# The package 'pcsets' is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# The package 'pcsets' is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty
# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# $Id: __init__.py 191 2007-08-07 03:15:46Z mccosar $

__all__ = """
pcset
pcops
catalog
noteops
tonerow
""".split()


pcsets/pcset.py
#!/usr/bin/env python

# pcsets 2.0.2 -- Pitch Class Sets for Python.
#
# Copyright 2007 Bruce H. McCosar
#
# This file is part of the package 'pcsets'
#
# The package 'pcsets' is free software; you can redistribute it
# and/or modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 3 of
# the License, or (at your option) any later version.
#
# The package 'pcsets' is distributed in the hope that it will be
# useful, but WITHOUT ANY WARRANTY; without even the implied warranty
# of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
# General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.
#
# $Id: pcset.py 191 2007-08-07 03:15:46Z mccosar $

"""
pcset.py -- version 2.0 -- Bruce H. McCosar

Pitch Class Sets are a mathematical model for analyzing and composing music.
Each note 'C' through 'B' has an equivalent pitch class number 0 through 11:

      | common |      C#      D#          F#      G#      A#      |
      |  note  |  C   --  D   --  E   F   --  G   --  A   --  B   |
      |  name  |      Db      Eb          Gb      Ab      Bb      |
      |        |                                                  |
      |  pc #  |  0   1   2   3   4   5   6   7   8   9   10  11  |

Sets of these numbers may be operated on by mathematical functions such as
transpose and invert, and analyzed by such measures as interval vectors and
relations to other sets, particularly prime sets.


Features:

  * All PcSet methods return NEW PcSets instead of modifying the original in
    place. This makes it possible to chain calls together, such as:

    newpcs = pcs.invert().transpose(4).reverse()

  * The basic PcSet class does not understand note names, only numerical pitch
    classes. Note operations are in a separate module(noteops.py).

    There are two ways to define pitch classes (PcSets):

    -- As a specificiation string.  Fast and easy, this is simply a string
       containing the numerical values for the pitch classes, with A = 10 and
       B = 11.  Example: PcSet('0146A')

    -- As a list of integers.  Example: [0,1,3,4,6,7,9,10]

    Duplicate values in the definition will be ignored.  All numbers will be
    mapped to the modulus 12 domain (from 0 to 11)

  * The PcSet class supports __len__, __str__, and __iter__, making the
    following calls possible (and replacing some of the version 1 methods):

    Purpose                                    NEW in v2.0  Old method
    =========================================  ===========  ==============
    Find cardinality of a set (# of elements)  len(pcs)     pcs.card()
    Return the set as a string                 str(pcs)     pcs.toString()
    Create a list of the elements in the set   list(pcs)    pcs.toList()


CORE as of version 2.0.0 -- new functionality may occasionally be added,
and bugs will be fixed, but the interface defined here will not change
in the entire version 2 series.
"""

__metaclass__ = type

__all__ = ('PcSet', 'PcSetException', 'DefinitionError')


class PcSetException(Exception):
    """
    The base class for exceptions in the pcset family of modules.
    """
    def __init__(self):
        self.message = self.__doc__

    def __str__(self):
        return self.message


class DefinitionError(PcSetException):
    """
    This exception type is thrown when something goes wrong in defining a
    PcSet (creating a PcSet instance). Any exception that occurs during
    definition will be a subclass of this.
    """


class IllegalCharacter(DefinitionError):
    """
    When defining a PcSet as a specification string, the only legal
    characters are the digits 0-9 and the capital letters A and B.
    The first illegal character encountered was: %(illegal)s
    """
    def __init__(self, ch):
        self.message = self.__doc__ % {'illegal': ch}


class NonIterableDef(DefinitionError):
    """
    When defining a PcSet, the definition must be an iterable type
    such as a list, set, tuple, or string. In this case, something
    non-iterable was encountered:

    problem input = %(problem)s %(classification)s
    """
    def __init__(self, x):
        trouble = {
            'problem': x,
            'classification': type(x)
            }
        self.message = self.__doc__ % trouble


def moderate(x):
    """
    A utility function. Returns the integer between 0 and 11 that best
    represents the input value mod 12. Character values for the digits 0
    through 9 are converted to integers; the characters 'A' and 'B' are
    interpreted as 10 and 11, respectively.
    """
    try:
        n = int(x)
    except ValueError:
        if x == 'A':
            n = 10
        elif x == 'B':
            n = 11
        else:
            raise IllegalCharacter(x)
    return n % 12


def binaryvalue(i):
    """
    A utility function. Finds the binary 'index value' for a list of integers,
    encoded so that if an integer is present in the list, it is one 'bit' in
    the binary number. Therefore, [0,1,7] would be 2**0 + 2**1 + 2**7.  Since
    PcSets are guaranteed to have no duplicates, this gives a unique integer
    between 0 and 4095 for each possible set.
    """
    value = 0
    for bit in i:
        value += 2**bit
    return value


class PcSet:

    """
    Pitch Class Set. Once defined, the set can be manipulated using the
    supplied methods, which are grouped into the following categories:

    FUNDAMENTAL METHODS
        invert()
        transpose(n)

    SET OPERATIONS
        complement()
        reverse()
        sort()
        shift(n)
        zero()
        normal()
        reduced()
        prime()

    SET ANALYSIS
        ivec()
        cvec()

    SHORTHAND METHODS
        I()
        T(n)
        TnI(n)
        Ixy(x,y)

    In addition, PcSets have default string, length, and iterator methods,
    making possible calls such as len(pcs), str(pcs), and list(pcs).

    A useful reference for understanding pc sets:
    http://www.jaytomlin.com/music/settheory/help.html
    """

    # basic services

    def __init__(self, definition):
        """
        A PcSet may be defined as a spec string or as a list. If a string is
        entered, the only characters allowed are 0-9 and A-B (A represents
        10 and B is 11). If a problem is encountered with the input to the
        constructor, a DefinitionError is raised.
        """
        try:
            redefinition = [moderate(note) for note in list(definition)]
        except TypeError:
            # non-iterables go here
            raise NonIterableDef(definition)
        self.definition = []
        for note in redefinition:
            # remove duplicates
            if note not in self.definition:
                self.definition.append(note)

    def __iter__(self):
        for note in self.definition:
            yield note

    def __str__(self):
        output = ''
        for x in self:
            if x < 10:
                output += str(x)
            elif x == 10:
                output += 'A'
            elif x == 11:
                output += 'B'
        return output

    def __repr__(self):
        return 'PcSet(%s)' % self.definition

    def __len__(self):
        return len(self.definition)

    # fundamental methods

    def invert(self):
        """
        Returns a new PcSet which is the inverse of the original.
        """
        inverse = [((12-note) % 12) for note in self]
        return PcSet(inverse)

    def transpose(self, n):
        """
        Returns a new PcSet which is the original transposed by n.
        """
        transposed = [((note+n) % 12) for note in self]
        return PcSet(transposed)

    # set operations

    def complement(self):
        """
        Returns a new PcSet which is the complement of the original set -- it
        contains all the elements which the original does not.
        """
        anti = []
        for note in range(12):
            if note not in self:
                anti.append(note)
        return PcSet(anti)

    def reverse(self):
        """
        Returns a new PcSet with the elements of the original reversed.
        """
        return PcSet(reversed(self.definition))

    def sort(self):
        """
        Returns a new PcSet in which the elements of the original have been
        sorted in ascending order.
        """
        return PcSet(sorted(self))

    def shift(self, n):
        """
        Returns a new PcSet in which the elements of the original have been
        shifted up 'n' places. (Negative values for n shift down instead; zero
        does nothing but return a copy.)
        """
        size = len(self)
        copy = list(self)
        if size > 1:
            uptimes = int(n) % size
            if uptimes > 0:
                for x in range(uptimes):
                    copy.insert(0, copy.pop())
        return PcSet(copy)

    def zero(self):
        """
        Returns a new PcSet in which the elements have been transposed so that
        the first element is zero.
        """
        try:
            return self.transpose(-self.definition[0])
        except IndexError:
            # empty set, no first element
            return PcSet(self)

    def normal(self):
        """
        Returns a new PcSet which is the 'normal form' of the original. The
        normal form has elements arranged to take up smallest total interval
        space from beginning to end. In the case of ties, the arrangement with
        the best 'packing' toward the left is chosen.
        """
        size = len(self)
        if size < 2:
            return PcSet(self)
        # determine all the possible shift rotations of the sorted set
        original = self.sort()
        rotations = [original.shift(n) for n in range(size)]
        # determine which arrangement has the lowest binary value
        # when compared from a fair start (first element on zero)
        bestnormal = rotations.pop()
        established = binaryvalue(bestnormal.zero())
        for arrangement in rotations:
            challenger = binaryvalue(arrangement.zero())
            if challenger < established:
                bestnormal = arrangement
                established = challenger
        return bestnormal

    def reduced(self):
        """
        Returns a new PcSet which is the 'reduced form' of the original.
        For a given PcSet 'pcs', this is equivalent to the operation series

            pcs.normal().zero()

        The returned set will be easily comparable to other reduced sets.
        For example:

        >>> from pcsets.noteops import pcfor
        >>> examples = [pcfor(s) for s in ('C E G', 'A C# E', 'A C E')]
        >>> print [str(p) for p in examples]
        ['047', '914', '904']
        >>> print [str(p.normal()) for p in examples]
        ['047', '914', '904']
        >>> print [str(p.reduced()) for p in examples]
        ['047', '047', '037']
        >>> print [str(p.prime()) for p in examples]
        ['037', '037', '037']

        The chords C major, A major, and A minor are not easily comparable in
        normal form. When the normal form is transposed to start on zero, it's
        clear that '047' must be the reduced form that corresponds to what we
        think of as 'major' chords.

        Note, however, that the difference disappears when these are put in
        prime form. (See the documentation for the prime() method.)
        """
        return self.normal().zero()

    def prime(self):
        """
        Returns a new PcSet which is the 'prime form' of the original.
        The prime form can be transformed into any of the sets in its
        'family' through the operations T(n) and T(n)I.

        Any set will have a normal form.  If the set is transposed so that
        the first element is zero, this is the reduced form.  However, any
        set will also have an inversion.  The reduced form of the inversion
        can be compared to the original reduced form.  The prime form,
        then, is the reduced set that has the closest leftward packing.
        """
        original = self.normal().zero()           # reduced form
        inverted = self.invert().normal().zero()  # reduced form of inverse
        if binaryvalue(original) < binaryvalue(inverted):
            return original
        else:
            return inverted

    # set analysis

    def ivec(self):
        """
        Finds the Interval Vector for the set. This is defined as a six member
        vector, with each value representing an interval group:

        [0] = Number of Group 1 inversions (minor 2nd, major 7th)
        [1] = Same, but for Group 2 (major 2nd, minor 7th)
        ...   (and so on)
        [5] = Same, but for Group 6 (the tritone)
        """
        ivec = [0] * 6
        workingcopy = list(self.sort())
        while len(workingcopy) > 1:
            note = workingcopy.pop()
            for othernote in workingcopy:
                intervalclass = (note - othernote) % 12
                if intervalclass > 6:
                    intervalclass = (12 - intervalclass) % 12
                ivec[intervalclass-1] += 1
        return ivec

    def cvec(self):
        """
        Common Tone Vector: finds the number of common tones for each possible
        value of n in the operation TnI. Returns a twelve member vector where
        element 0 is the number of common tones under T(0)I, 1 under T(1)I,
        and so on.
        """
        cvec = [0] * 12
        rawtable = []
        for x in self:
            for y in self:
                rawtable.append((x+y) % 12)
        entries = set(rawtable)
        for value in entries:
            cvec[value] += rawtable.count(value)
        return cvec

    # shorthand methods

    def I(self):
        """
        Shorthand for invert()
        """
        return self.invert()

    def T(self, n):
        """
        Shorthand for transpose(n)
        """
        return self.transpose(n)

    def TnI(self, n):
        """
        Shorthand for two combined operations: first inversion, then
        transposition by n. (Although class methods use postfix notation, the
        original pc set theory literature commonly uses prefix notation such
        as this.)
        """
        return self.invert().transpose(n)

    def Ixy(self, x, y):
        """
        Shorthand for inversion around an axis specified by pitch classes x
        and y. Inversion around this axis will cause x to become y and y to
        become x.
        """
        return self.invert().transpose(x+y)
